modules/aidemBidAdapter.js

```diff
@@ -0,0 +1,691 @@
+      deepSetValue(notice.metrics, 'adapter.client.buildRequests', metrics['adapter.client.buildRequests'])
+      deepSetValue(notice.metrics, 'adapter.client.total', metrics['adapter.client.total'])
+      deepSetValue(notice.metrics, 'adapter.client.net', metrics['adapter.client.net'])
+      deepSetValue(notice.metrics, 'adapter.client.interpretResponse', metrics['adapter.client.interpretResponse'])
+    }
+
+    timeoutNotice.bids.push(notice)
+  })
+
+  return timeoutNotice
+}
+
+function buildErrorNotice(prebidErrorResponse) {
+  const errorNotice = {
+    message: `Prebid.js: Server call for ${prebidErrorResponse.bidderCode} failed.`,
+    url: encodeURIComponent(getPageUrl(prebidErrorResponse)),
+    auctionId: prebidErrorResponse.auctionId,
+    bidderRequestId: prebidErrorResponse.bidderRequestId,
+    metrics: {}
+  }
+  if (prebidErrorResponse.metrics && typeof prebidErrorResponse.metrics.getMetrics === 'function') {
+    const metrics = prebidErrorResponse.metrics.getMetrics()
+    deepSetValue(errorNotice.metrics, 'requestBids.validate', metrics['requestBids.validate'])
+    deepSetValue(errorNotice.metrics, 'requestBids.makeRequests', metrics['requestBids.makeRequests'])
+    deepSetValue(errorNotice.metrics, 'requestBids.total', metrics['requestBids.total'])
+    deepSetValue(errorNotice.metrics, 'requestBids.callBids', metrics['requestBids.callBids'])
+    deepSetValue(errorNotice.metrics, 'adapter.client.validate', metrics['adapter.client.validate'])
+    deepSetValue(errorNotice.metrics, 'adapter.client.buildRequests', metrics['adapter.client.buildRequests'])
+    deepSetValue(errorNotice.metrics, 'adapter.client.total', metrics['adapter.client.total'])
+    deepSetValue(errorNotice.metrics, 'adapter.client.net', metrics['adapter.client.net'])
+    deepSetValue(errorNotice.metrics, 'adapter.client.interpretResponse', metrics['adapter.client.interpretResponse'])
+  }
+  return errorNotice
+}
+
+/**
+ * Get One size from Size Array
+ * [[250,350]] -> [250, 350]
+ * [250, 350]  -> [250, 350]
+ * @param {array} sizes array of sizes
+ */
+function getFirstSize(sizes = []) {
+  if (isValidSize(sizes)) {
+    return sizes;
+  } else if (isValidSize(sizes[0])) {
+    return sizes[0];
+  }
+
+  return false;
+}
+
+function hasValidFloor(obj) {
+  if (!obj) return false
+  const hasValue = !isNaN(Number(obj.value))
+  const hasCurrency = contains(AVAILABLE_CURRENCIES, obj.currency)
+  return hasValue && hasCurrency
+}
+
+function getMediaType(bidRequest) {
+  if ((bidRequest.mediaTypes && bidRequest.mediaTypes.hasOwnProperty('video')) || bidRequest.params.hasOwnProperty('video')) { return VIDEO }
+  return BANNER
+}
+
+function getPrebidRequestFields(bidderRequest, bidRequests) {
+  const payload = {}
+  // Base Payload Data
+  deepSetValue(payload, 'id', bidderRequest.auctionId);
+  // Impressions
+  setPrebidImpressionObject(bidRequests, payload)
+  // Device
+  deepSetValue(payload, 'device', getDevice())
```

Comment:
For adapters that speak ORTB you get almost everything for free if you use our [ORTB conversion library](https://github.com/prebid/Prebid.js/blob/master/libraries/ortbConverter/README.md).

Reply:
@dgirardi we prefer to manually build the outgoing payload.

Reply:
Please add detail or use the library

Reply:
@patmmccann our backend is built with Rust, and we enforce strict type checking across all incoming payloads' data structures (BidRequest/ErrorNotice/WinNotice). While this allow us to have a more robust SSP, the flipside is that any unexpected change on the Prebid payloads will error out on the SSP.
This is why we have manually structured the Prebid payloads, and why in our test suite we check the types (and we run our own BrowserStack CI across dozens of browser/device combo). The ORTB conversion library doesn't allow us to enforce the type structure (the library might change, or different browsers/device might have different types (e.g., “true” instead of “1”).